src/main/scala/firrtl/passes/RemoveAccesses.scala

```diff
@@ -14,92 +14,90 @@ import scala.collection.mutable
 object RemoveAccesses extends Pass {
   def name = "Remove Accesses"
 
-  /** Container for a base expression and its corresponding guard
-    */
-  case class Location(base: Expression, guard: Expression)
+  def run(c: Circuit): Circuit = {
+    // This improves the performance of this pass
+    val createExpsCache = mutable.HashMap[Expression, Seq[Expression]]()
+    def create_exps(e: Expression) =
+      createExpsCache getOrElseUpdate (e, firrtl.Utils.create_exps(e))
+
+    /** Container for a base expression and its corresponding guard
+      */
+    case class Location(base: Expression, guard: Expression)
 
-  /** Walks a referencing expression and returns a list of valid references
-    * (base) and the corresponding guard which, if true, returns that base.
-    * E.g. if called on a[i] where a: UInt[2], we would return:
-    *   Seq(Location(a[0], UIntLiteral(0)), Location(a[1], UIntLiteral(1)))
-    */
-  def getLocations(e: Expression): Seq[Location] = e match {
-    case e: WRef => create_exps(e).map(Location(_,one))
-    case e: WSubIndex =>
-      val ls = getLocations(e.exp)
-      val start = get_point(e)
-      val end = start + get_size(tpe(e))
-      val stride = get_size(tpe(e.exp))
-      val lsx = mutable.ArrayBuffer[Location]()
-      for (i <- 0 until ls.size) {
-        if (((i % stride) >= start) & ((i % stride) < end)) {
-          lsx += ls(i)
+    /** Walks a referencing expression and returns a list of valid references
+      * (base) and the corresponding guard which, if true, returns that base.
+      * E.g. if called on a[i] where a: UInt[2], we would return:
+      *   Seq(Location(a[0], UIntLiteral(0)), Location(a[1], UIntLiteral(1)))
+      */
+    def getLocations(e: Expression): Seq[Location] = e match {
+      case e: WRef => create_exps(e).map(Location(_,one))
+      case e: WSubIndex =>
+        val ls = getLocations(e.exp)
+        val start = get_point(e)
+        val end = start + get_size(tpe(e))
+        val stride = get_size(tpe(e.exp))
+        for ((l, i) <- ls.zipWithIndex
+          if ((i % stride) >= start) & ((i % stride) < end)) yield l
+      case e: WSubField =>
+        val ls = getLocations(e.exp)
+        val start = get_point(e)
+        val end = start + get_size(tpe(e))
+        val stride = get_size(tpe(e.exp))
```

Comment:
Related to the commends about getting rid of the tpe function, at the very least new uses should use the class method Expression.tpe since the virtual function call will be faster than matching.